sellercenter-attributes/src/main/java/com/gfg/sellercenter/attributes/service/AttributeService.java

```diff
@@ -0,0 +1,78 @@
+import com.gfg.sellercenter.attributes.entity.AttributeWithOption;
+import com.gfg.sellercenter.attributes.entity.ProductWithAttributeInformation;
+import com.gfg.sellercenter.attributes.reader.HttpReader;
+
+import lombok.AllArgsConstructor;
+
+import org.json.JSONArray;
+import org.json.JSONObject;
+
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.stream.Collectors;
+
+@AllArgsConstructor
+public class AttributeService {
+
+    private final HttpReader httpReader;
+
+    public Map<Integer, ProductWithAttributeInformation>
+            getSerialNumberRequiredAttributesByProductIds(List<Integer> productIds)
+                    throws Exception {
+        productIds = productIds.stream().distinct().collect(Collectors.toList());
+        return createAttributeEntity(
+                this.httpReader.getSerialNumberRequiredAttributesByProductIds(productIds));
+    }
+
+    private Map<Integer, ProductWithAttributeInformation> createAttributeEntity(
+            JSONArray responseArray) {
+        Map<Integer, ProductWithAttributeInformation> categoryToProductMap = new HashMap<>();
+
+        for (Object response : responseArray) {
+            JSONObject resp = (JSONObject) response;
+            ProductWithAttributeInformation productWithAttribute =
+                    categoryToProductMap.get(resp.getInt("id"));
+
+            if (null == productWithAttribute) {
+                Map<Integer, AttributeWithOption> attributeWithOption = new HashMap<>();
+                categoryToProductMap.put(
+                        resp.getInt("id"),
+                        new ProductWithAttributeInformation(
+                                resp.getInt("id"),
+                                resp.getString("skuSeller"),
+                                resp.getJSONObject("attributes"),
+                                attributeWithOption));
+            }
+            this.addAttributeWithOption(
+                    categoryToProductMap.get(resp.getInt("id")),
+                    resp.getJSONObject("attributeWithOption"));
+        }
+
+        return categoryToProductMap;
+    }
+    private void addAttributeWithOption(ProductWithAttributeInformation product, JSONObject response) {
+
+        Integer attributeId = response.getInt("id");
+        AttributeWithOption attributeWithOption = product.getAttribute().get(attributeId);
+
+        if (null == attributeWithOption) {
+            product.getAttribute()
+                    .put(
+                            attributeId,
+                            new AttributeWithOption(
+                                    attributeId, response.getString("label"), new HashMap<>()));
+        }
+        product.getAttribute()
+                .get(attributeId)
+                .getAttributeOptions()
+                .put(
+                        response.getJSONObject("attributeOption").getInt("id"),
+                        new AttributeOption(
```

Comment:
I know that you like more to keep everything under control and not use deserializing using Jackson, but doing so will shrink current code. 
Basically the whole code can be removed and it will be something like
```
ProductWithAttributeInformation[] array = mapper.readValue(hereStringWithResponse, ProductWithAttributeInformation[].class);
// code to convert plain array or list to map, something like (write in browser, but you get my idea)
Map<Integer, ProductWithAttributeInformation> map = new HashMap<>();
for (ProductWithAttributeInformation item : array) {
  map.putIfAbsent(item.getId(), item);
  map.get(item.getId()). getAttributeOptions().put(...)
}
```

Reply:
Also for other reviewers: Jackson default mapper can be used to convert response to list of some objects (not the final objects we need). And then this list will have to be reduced/mapped to objects that we really need (roughly: take all the records for one product, create ProductWithAttributeInformation, add attributes to that object).
So this approach will reduce complexity in one place but will require another portion of code which will be quite similar to what we have now. 
Therefore now I support Dennis with keeping whole mapping as it is now.